features/101-design-webhooks/design.md

```diff
@@ -0,0 +1,156 @@
+
+For some events, it may allow to webhook to modify changes (e.x. auth gear's before family hooks) when response code is 2xx. 
+
+## Configure webhook
+
+A gear should provide an REST interface for configuring webhooks, it should follow following format:
+
+- create hooks
+
+```bash=
+curl -X POST -H "Content-Type: application/json" \
+-d @- http://localhost:3000/<gear name>/hooks <<EOF
+[
+    {
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "secret": "secret_string"
+    },
+    {
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "secret": "secret_string"
+    },
+]
+EOF
+[
+    {
+        "id": "<hook id>",
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "updated_at": "2019-01-06T20:39:23Z",
+        "created_at": "2019-01-06T17:26:27Z"
+    },
+    {
+        "id": "<hook id>",
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "updated_at": "2019-01-06T20:39:23Z",
+        "created_at": "2019-01-06T17:26:27Z"
+    },
+]
+```
+
+- Get hooks
+```bash=
+curl http://localhost:3000/<gear name>/hooks
+EOF
+[
+    {
+        "id": "<hook id>",
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "updated_at": "2019-01-06T20:39:23Z",
+        "created_at": "2019-01-06T17:26:27Z"
+    },
+    {
+        "id": "<hook id>",
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "updated_at": "2019-01-06T20:39:23Z",
+        "created_at": "2019-01-06T17:26:27Z"
+    },
+]
+```
+
+- Modify a hook
+```bash=
+curl -X PATCH -H "Content-Type: application/json" \
+-d @- http://localhost:3000/<gear name>/hooks/<hook id> <<EOF
+{
+    "events": ["<name of the event>", "<name of the event>"],
+    "url": "<url of the endpoint>",
+    "async": true|false,
+    "secret": "secret_string"
+}
+EOF
+{
+    "id": "<hook id>",
+    "events": ["<name of the event>", "<name of the event>"],
+    "url": "<url of the endpoint>",
+    "async": true|false,
+    "updated_at": "2019-01-06T20:39:23Z",
+    "created_at": "2019-01-06T17:26:27Z"
+}
+```
+
+- Delete a hook
+```bash=
+curl -X DELETE -H "Content-Type: application/json" \
+http://localhost:3000/<gear name>/hooks/<hook id>/delete 
+EOF
+{
+     "result": "OK"
+}
+```
+
+## Verify Request
+
+`X-Skygear-Webhhok-Signature` is generated the HMAC hex digest value from the `secret` of a hook.
+
+## TBD
+
+- How `X-Skygear-Webhhok-Signature` works?
+- Should webhook provide retry mechanism?
```

Comment:
Sure. The most standard way is to respect 503 Services Unavailable or 429 Too Many Requests, if the `Retry-After` header was set.  Similarly, we need to consider if we shall respect 301, 302, 307, 308 for redirection,

Reply:
1. For `Retry-After`, I have concerns if it returns a very long delay, it may cause problems on `SYNC` webhooks.
2. For 3xx redirect response, I have concerns if infinite loop redirection happens, so I would like to consider it is a failed response.

Reply:
1. Sorry didn't make it clear, I think SYNC request would have to ignore `Retry-After`
1. For redirection, we can simply like a browser to have a counter of max redirect. A practical use case is a redirection to a new version, and people can track the old hooks via logs.

But I agree both features are not needed in first launch (but on the spec we should have it and open issues later)